File: Programmers/codingTest_Kit/Stack_Queue/42583.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    
    int idx = 0, sum = 0;
    queue<int> q;

    while(1){
        if(idx == truck_weights.size()){
            answer += bridge_length;
            break;
        }
        answer++;
        int tmp = truck_weights[idx];
        if(q.size() == bridge_length){
            sum -= q.front();
            q.pop();
        }
        if(sum + tmp <= weight){
            q.push(tmp);
            sum += tmp;
            idx++;
        }else{
            q.push(0);
        }
    }

    return answer;
}

File: Programmers/Practice/LV2/42583.java
package Programmers.Practice.LV2;

import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int sum = 0;
        int count = 0;
        Queue<Integer> bridgeQue = new ArrayDeque<>();
        Queue<Integer> startQue = new ArrayDeque<>();
        Queue<Integer> endQue = new ArrayDeque<>();

        for (int i = 0; i < truck_weights.length; i++) {
            startQue.add(truck_weights[i]);
        }

        for (int i = 0; i < bridge_length; i++) {
            bridgeQue.add(0);
        }

        while (endQue.size() != truck_weights.length) {
            answer++;
            if (bridgeQue.size() == bridge_length) {
                if (bridgeQue.peek() != 0) {
                    endQue.add(bridgeQue.peek());
                }
                sum -= bridgeQue.poll();
            }

            if (!startQue.isEmpty() && startQue.peek() + sum <= weight) {
                bridgeQue.add(startQue.peek());
                sum += startQue.poll();
            } else {
                bridgeQue.add(0);
            }

        }

        return answer;
    }
}